### PR TITLE
Optimize Multi-Point Evaluation & Parallelization Switch & Misc

### LINKED ISSUE #10: Allow Mesh Caching
At the moment some overhead is caused by re-initialization of variables between successive calls to the model. This could be remedied by allowing some functionality to cache some variables (such as mesh and normals) between calls.

### LINKED ISSUE #13: Allow to chose parallel or serial execution at runtime
The current implementation is compiled with either `CPP` (serial), `OMP` or `TBB` backend.

For E.g. the package on `conda` is always parallelized with `OMP`. This needs to change.

Thus, enable the user to choose `serial` or `parallel` execution at runtime via an argument given to the evaluated parameter. I.e. make the `CPP` backend always usable.

### PR DESCRIPTION
**Pull request by schuhmaj**
# Changelog

- Closes #10 
- Closes #13

## Build System & Overall Structure

- All sources in one place - Move everything into the source folder, including the Python Interface
- Restructure the `CMake` Build System - One target per `CMakeLists.txt`
- Bump up the version of `thrust` to 2.1
- Bump up the version of `GoogleTest` to 1.13
- Bump up the version of `pybind11` to 2.10.4
- Bump up the version of `xsimd` to 11.1.0 & Adapt internal components to new API

## C++

- Create `GravityEvaluable` class
  - The class caches properties like normals and vectors, which do not differ given different computation points
  - These properties a computed once at initialization
  - The public interface of the class is fully exposed to Python, including the syntactic sugar `operator()`
  - Add template parameter to choose parallelization backend
  - Serial & Parallel Methods are instantiated. Thus, the parallelization selection is made available via `operator()` at runtime
  - Make use of `std::variant` to further reduce the code footprint
- Existing `GravityModel` namespace
  - Redirect the existing functions' calls to the new class to avoid code duplication and sustain compatibility with the existing API
  - Add parallelization parameter
  - Get rid of `GravityModelResult` and use a standard `tuple` (facilitates the Python Interface)
  - Remove the "rounding error" method. I argue `1e-26` is as good as `0.0`. Postprocessing can be done by the user if wished
  - Make use of `std::variant` to further reduce the code footprint
- `Util` component
  - Add template method for adding the contents of two `std::tuple`
- Adapt tests to the changed return value

## Python

- Reduced the code footprint of the interface declaration by making use of `std::variant`
- New parameter: `parallel` to enable/ disable parallelization
- New Interface for `evaluate`

```python
import polyhedral_gravity as model

# The "free function" evaluate has now always three arguments (or an optional fourth one)
# The keywords of the signature are now always the SAME!
# So the third argument takes computation_points or a SINGLE computation_point
# Optionally with the keyword prefix
# The "parallel" argument allows to disable the parallelization
model.evaluate(polyhedral_source=(vertices, faces), density=1.0, computation_points=[..], parallel=True)

# Removed signature (vertices and faces as separate arguments)
# model.evaluate(vertices, faces, density=1.0, computation_points=[..])

model.evaluate(polyhedral_source=["files",...], density=1.0, computation_points=[..], parallel=True)

```

- New class `GravityEvaluable`

```python
import polyhedral_gravity as model

# Construction like above, but only with the two "persistent" arguments
evaluable = polyhedral_gravity.GravityEvaluable((vertices, faces), density)

for i in range(N):
   # Second argument is optional
    evaluable(computation_points[i], parallel=True)

```

Results from comparing the old & new interface (N=1, so small runtime differences are not really expressive)
1000 random sampled points around Eros using Apple Clang 14, M1 Pro and the TBB Backend.

```
######## Old Single-Point ########
--> 1000 times 1 point with old interface
--> Time taken: 7765.073 microseconds per point
######## Old Multi-Point ########
--> 1 time N points with old interface
--> Time taken: 275.917 microseconds per point
######## NEW GravityEvaluable (Single-Point) #########
--> 1000 times 1 point with GravityEvaluable
--> Time taken: 313.408 microseconds per point
######## NEW GravityEvaluable (Multi-Point) ########
--> 1 time 1000 points with GravityEvaluable
--> Time taken: 253.031 microseconds per point
```

==> Basically, there is no reason to use the old interface, as the new one only provides advantages

## Docs

- Updated the docs to reflect the new changes

### CODE REVIEW

**Comment by schuhmaj** [comment]
Open TODOs (after/ before merge)

- [ ] Publish on PyPi/ Add a CI that does this
- [ ] Update Conda Recipe

**Comment by schuhmaj** [comment]
@gomezzz The public API changed slightly. So old calls won't necessarily work (`vertices` and `faces` are no handed over as a tuple (in C++ and Python) & the C++ return type changed (not the Python one!)). Should we thus either go from `1.2.x` to

- `1.3` (because there are only minor API changes) <-- Current Set-Up in the PR
- `2.0` (because some calls are refactored/ old ones not working anymore)

**Comment by gomezzz** [comment]
> @gomezzz The public API changed slightly. So old calls won't necessarily work (`vertices` and `faces` are no handed over as a tuple (in C++ and Python) & the C++ return type changed (not the Python one!)). Should we thus either go from `1.2.x` to
> 
>     * `1.3` (because there are only minor API changes) <-- Current Set-Up in the PR
> 
>     * `2.0` (because some calls are refactored/ old ones not working anymore)

Yea it might be nice to make it 2.0, I guess.

In regard to the API changes, did / could you include warnings/errors to the user to make sure they are aware of the API change? Always easier to tell people in case they update and suddenly their code stops working 🙈

File: docs/quick_start_python.rst [hunk 1]
```diff
@@ -2,7 +2,34 @@ Quick Start Python
...
+    #############################################
+    # Define the polyhedral_source which is either (vertices, faces) or a list of files
+    # Define the density (a float)
+    # Define the computation_points which is either a single point or a list of points
+    # Define if the computation is parallel or not (the default is parallel, which corresponds to True)
+    results = model.evaluate(polyhedral_source, density, computation_point, parallel)
+
+    #############################################
+    # With the evaluable class
+    #############################################
```
Comment (gomezzz) [inline docs/quick_start_python.rst:25]:
```suggestion
    # With the evaluable class
    # This allows multiple evaluations on the same polyhedron
    # without the overhead of setting up the normals etc.
    #############################################
```

File: docs/quick_start_python.rst [hunk 2]
```diff
@@ -2,7 +2,34 @@ Quick Start Python
...
+    # Parameters are the same as for the free function call
+    evaluable = model.GravityEvaluable(polyhedral_source, density)
```
Comment (gomezzz) [inline docs/quick_start_python.rst:27]: If possible I would suggest the use of [named argument, helps readability](https://trstringer.com/python-named-arguments/#self-documenting-code)

File: docs/quick_start_python.rst [hunk 3]
```diff
@@ -18,13 +45,14 @@ defined from within source code for a specific point and density.
         computation_point = ... # [] or np.array of length 3
 
         # Evaluate the gravity model
-        # Notice that the last argument could also be a list of points
-        # Returns (list of) tuple of potential, acceleration and tensor
-        potential, acceleration, tensor = model.evaluate(vertices, faces, density, computation_point)
+        # Notice that the third argument could also be a list of points
+        # Returns a tuple of potential, acceleration and tensor
+        potential, acceleration, tensor = model.evaluate((vertices, faces), density, computation_point, parallel=True)
```
Comment (gomezzz) [inline docs/quick_start_python.rst:50]: named arguments? :)

File: src/main.cpp [hunk 4]
```diff
@@ -10,8 +10,8 @@ int main(int argc, char *argv[]) {
     using namespace polyhedralGravity;
     if (argc != 2) {
         SPDLOG_LOGGER_INFO(PolyhedralGravityLogger::DEFAULT_LOGGER.getLogger(), "Wrong program call! "
-                                                                                 "Please use the program like that:\n"
-                                                                                 "./polyhedralGravity [YAML-Configuration-File]\n");
+                                                                                "Please use the program like that:\n"
```
Comment (gomezzz) [inline src/main.cpp:13]:
```suggestion
                                                                                "Please use the program like this:\n"
```

**Review by gomezzz — CHANGES_REQUESTED** [review]
some minor comments

File: cmake/gtest.cmake [hunk 5]
```diff
@@ -3,10 +3,10 @@ include(FetchContent)
 message(STATUS "Setting up gtest")
 
 #Adapted from https://cliutils.gitlab.io/modern-cmake/chapters/testing/googletest.html
-#Fetches the version 1.11.0 from the official github for googletest
+#Fetches the version 1.13.0 from the official github for googletest
```
[the remark this answers is not in the document]
Comment (schuhmaj) [reply cmake/gtest.cmake:6]: We skipped that :)

File: docs/quick_start_python.rst (hunk 3 above)
Comment (schuhmaj) [reply docs/quick_start_python.rst:52]: Yes, I use both notations, so the reader can basically chose the flavour :)

File: src/polyhedralGravity/util/UtilityContainer.h [hunk 6]
```diff
@@ -327,6 +327,35 @@ namespace polyhedralGravity::util {
         return std::abs(x - y) > maxExponentDifference;
     }
 
+    namespace detail {
```
[the remark this answers is not in the document]
Comment (schuhmaj) [reply src/polyhedralGravity/util/UtilityContainer.h:330]: I would keep it named `detail` boost, and also other C++ libraries uses `detail`  to mark private namespaces, which should not be used from the outside.

File: README.md [hunk 7]
```diff
@@ -38,13 +38,13 @@ on [readthedocs](https://polyhedral-gravity-model-cpp.readthedocs.io/en/latest/)
...
+- GoogleTest 1.13.0 (only required for testing)
 - spdlog 1.11.0 (required for logging)
 - tetgen 1.6 (required for I/O)
 - yaml-cpp 0.7.0 (required for I/O)
-- thrust 1.16.0 (required for parallelization and utility)
-- xsimd 8.1.0 (required for vectorization of the `atan(..)`)
-- pybind11 2.10.3 (required for the Python interface, but not the C++ standalone)
+- thrust 2.1.0 (required for parallelization and utility)
+- xsimd 11.1.0 (required for vectorization of the `atan(..)`)
+- pybind11 2.10.4 (required for the Python interface, but not the C++ standalone)
```
[the remark this answers is not in the document]
Comment (schuhmaj) [reply README.md:47]: Good point

**Comment by schuhmaj** [comment]
Automatic CI will be on a separate merge request, I will first publish the package on PyPi manually once.

@gomezzz I packed the dependencies' license links inside a note in our project's license. This is the way how our dependencies to it.

**Review by gomezzz — APPROVED** [review]
Looks good! For pypi, I would strongly recommend trying with testpypi first :v:

If it helps with automation, here are some workflows I use for uploading to both

https://github.com/esa/torchquad/blob/main/.github/workflows/deploy_to_test_pypi.yml
https://github.com/esa/torchquad/blob/main/.github/workflows/deploy_to_pypi.yml

**Pull request merged**